File: module-1/proof_of_work.py
import hashlib
import time

class Block:
    def __init__(self, index, previous_hash, timestamp, data, nonce=0):
        # Initialize a new block with the given parameters
        self.index = index
        self.previous_hash = previous_hash
        self.timestamp = timestamp
        self.data = data
        self.nonce = nonce
        self.hash = self.calculate_hash()  # Calculate and set the hash

    def calculate_hash(self):
        # Calculate the hash of the block's contents using SHA-256
        data = str(self.index) + self.previous_hash + str(self.timestamp) + self.data + str(self.nonce)
        return hashlib.sha256(data.encode()).hexdigest()

    def mine_block(self, difficulty):
        # Mine the block by finding a nonce that produces a hash with a prefix of zeros
        target = '0' * difficulty
        while self.hash[:difficulty] != target:
            self.nonce += 1
            self.hash = self.calculate_hash()

    def __str__(self):
        # String representation of the block for easy printing
        return f"Block {self.index} - Timestamp: {self.timestamp}, Nonce: {self.nonce}\nData: {self.data}\nHash: {self.hash}\nPrevious Hash: {self.previous_hash}"

class Blockchain:
    def __init__(self):
        # Initialize the blockchain with a genesis block
        self.chain = [self.create_genesis_block()]
        self.difficulty = 4  # Adjust the difficulty as needed

    def create_genesis_block(self):
        # Create the genesis block with predefined values
        return Block(0, "0", int(time.time()), "Genesis Block")

    def get_latest_block(self):
        # Get the latest block in the blockchain
        return self.chain[-1]

    def add_block(self, new_block):
        # Add a new block to the blockchain
        new_block.index = len(self.chain)
        new_block.previous_hash = self.get_latest_block().hash
        new_block.timestamp = int(time.time())
        new_block.mine_block(self.difficulty)
        self.chain.append(new_block)

    def is_chain_valid(self):
        # Check if the blockchain is valid by verifying hashes and previous hash links
        for i in range(1, len(self.chain)):
            current_block = self.chain[i]
            previous_block = self.chain[i - 1]

            if current_block.hash != current_block.calculate_hash():
                return False

            if current_block.previous_hash != previous_block.hash:
                return False

        return True

# Create a blockchain and add some blocks
my_blockchain = Blockchain()
my_blockchain.add_block(Block(1, my_blockchain.get_latest_block().hash, 0, "Transaction 1"))  # Pass the previous hash correctly
my_blockchain.add_block(Block(2, my_blockchain.get_latest_block().hash, 0, "Transaction 2"))  # Pass the previous hash correctly

# Display the blockchain
for block in my_blockchain.chain:
    print(block)
    print()

# Check if the blockchain is valid
print("Is blockchain valid?", my_blockchain.is_chain_valid())

File: module-1/proof_of_work.js
const crypto = require('crypto');

class Block {
    constructor(index, previousHash, timestamp, data, nonce = 0) {
        // Initialize a new block with the given parameters
        this.index = index;
        this.previousHash = previousHash;
        this.timestamp = timestamp;
        this.data = data;
        this.nonce = nonce;
        this.hash = this.calculateHash(); // Calculate and set the hash
    }

    calculateHash() {
        // Calculate the hash of the block's contents using SHA-256
        const data = `${this.index}${this.previousHash}${this.timestamp}${this.data}${this.nonce}`;
        return crypto.createHash('sha256').update(data).digest('hex');
    }

    mineBlock(difficulty) {
        // Mine the block by finding a nonce that produces a hash with a prefix of zeros
        const target = '0'.repeat(difficulty);
        while (this.hash.substring(0, difficulty) !== target) {
            this.nonce++;
            this.hash = this.calculateHash();
        }
        console.log(`Block mined: ${this.hash}`);
    }
}

class Blockchain {
    constructor() {
        // Initialize the blockchain with a genesis block
        this.chain = [this.createGenesisBlock()];
        this.difficulty = 4; // Adjust the difficulty as needed
    }

    createGenesisBlock() {
        // Create the genesis block with predefined values
        return new Block(0, '0', Date.now(), 'Genesis Block');
    }

    getLatestBlock() {
        // Get the latest block in the blockchain
        return this.chain[this.chain.length - 1];
    }

    addBlock(newBlock) {
        // Add a new block to the blockchain
        newBlock.index = this.chain.length;
        newBlock.previousHash = this.getLatestBlock().hash;
        newBlock.timestamp = Date.now();
        newBlock.mineBlock(this.difficulty);
        this.chain.push(newBlock);
    }

    isChainValid() {
        // Check if the blockchain is valid by verifying hashes and previous hash links
        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            if (currentBlock.hash !== currentBlock.calculateHash()) {
                return false;
            }

            if (currentBlock.previousHash !== previousBlock.hash) {
                return false;
            }
        }
        return true;
    }
}

// Create a blockchain and add some blocks
const myBlockchain = new Blockchain();
myBlockchain.addBlock(new Block(1, '', 0, 'Transaction 1'));
myBlockchain.addBlock(new Block(2, '', 0, 'Transaction 2'));

// Display the blockchain
myBlockchain.chain.forEach((block) => {
    console.log(block);
    console.log();
});

// Check if the blockchain is valid
console.log('Is blockchain valid?', myBlockchain.isChainValid());
